diffrax/_adjoint.py

```diff
@@ -852,3 +857,206 @@ def loop(
+        # Can we avoid this re-evaluation?
+
+        f0 = solver.func(terms, t1, y1, args)
+        step_y1, vjp_fun_y1 = eqx.filter_vjp(solver_step, terms, t1, t0, y1, args)
+        z0 = (ω(z1) - ω(y1) + ω(step_y1)).ω
+
+        f0 = solver.func(terms, t0, z0, args)
+        step_z0, vjp_fun_z0 = eqx.filter_vjp(solver_step, terms, t0, t1, z0, args)
+
+        y0 = ((1 / solver.l) * (ω(y1) - ω(step_z0)) + ω(z0)).ω
+
+        grad_step_y1 = vjp_fun_y1(grad_z1)
+        grad_y1 = (ω(grad_y1) + ω(grad_z1) - ω(grad_step_y1[3])).ω
+
+        grad_step_z0 = vjp_fun_z0(grad_y1)
+        grad_y0 = (solver.l * ω(grad_y1)).ω
+        grad_z0 = (ω(grad_z1) - solver.l * ω(grad_y1) + ω(grad_step_z0[3])).ω
+
+        grad_terms = (ω(grad_terms) - ω(grad_step_y1[0]) + ω(grad_step_z0[0])).ω
+        grad_args = (ω(grad_args) - ω(grad_step_y1[4]) + ω(grad_step_z0[4])).ω
+
+        return (
+            ts_index,
+            y0,
+            ((first_step, f0), z0),
+            grad_y0,
+            grad_z0,
+            grad_args,
+            grad_terms,
+        )
+
+    def cond_fun(state):
+        ts_index = state[0]
+        return ts_index > 0
+
+    state = (
+        ts_final_index,
+        y1,
+        (original_solver_state, z1),
+        grad_y1,
+        grad_z1,
+        grad_args,
+        grad_terms,
+    )
+
+    state = eqxi.while_loop(cond_fun, grad_step, state, kind="lax")
+    _, _, _, grad_y0, grad_z0, grad_args, grad_terms = state
+    return (ω(grad_y0) + ω(grad_z0)).ω, grad_args, grad_terms
+
+
+class ReversibleAdjoint(AbstractAdjoint):
+    """
+    Backpropagate through [`diffrax.diffeqsolve`][] when using the
+    [`diffrax.Reversible`][] solver.
+
+    This method implies very low memory usage and exact gradient calculation (up to
+    floating point errors).
+
+    This will compute gradients with respect to the `terms`, `y0` and `args` arguments
+    passed to [`diffrax.diffeqsolve`][]. If you attempt to compute gradients with
+    respect to anything else (for example `t0`, or arguments passed via closure), then
+    a `CustomVJPException` will be raised. See also
+    [this FAQ](../../further_details/faq/#im-getting-a-customvjpexception)
+    entry.
+    """
+
+    def loop(
+        self,
+        *,
+        args,
+        terms,
+        solver,
+        saveat,
+        init_state,
+        passed_solver_state,
+        passed_controller_state,
+        event,
+        **kwargs,
+    ):
+        # `is` check because this may return a Tracer from SaveAt(ts=<array>)
+        if eqx.tree_equal(saveat, SaveAt(t1=True)) is not True:
+            raise ValueError(
+                "Can only use `adjoint=ReversibleAdjoint()` with "
+                "`saveat=SaveAt(t1=True)`."
+            )
+
+        if not isinstance(solver, Reversible):
+            raise ValueError(
+                "Can only use `adjoint=ReversibleAdjoint()` with "
+                "`Reversible()` solver."
+            )
```

Comment:
Could we perhaps remove `Reversible` from the public API altogether, and just have `solver = Reversible(solver)` here? Make the `Reversible` solver an implementation detail of the adjoint.

Reply:
I really like this idea :D

I've removed `Reversible` from the public API and any `AbstractSolver` passed to `ReversibleAdjoint` is auto-wrapped. There is now a `_Reversible` class within the `_adjoint` module that is exclusively used by `ReversibleAdjoint`. Do you think this is an appropriate home for the `_Reversible` class or should I keep it elsewhere?